src/lib.rs

```diff
@@ -21,149 +75,262 @@
-//! generate many types; the [`Standard`] distribution carries the
-//! implementations. In some cases type annotation is required, e.g.
-//! `rng.gen::<f64>()`.
 //!
-//! # Getting random values
+//! ## Random number generators (RNGs)
+//!
+//! Pulling random numbers out of 'thin air' is very hard for a computer. Not
+//! impossible, but hard and slow. Generally the operating systems collects some
+//! of them, by measuring different kinds of noice/jitter in the hardware.
+//!
+//! What is always used instead is a psuedo-random number generator. This is a
+//! clever algorithm that can produce an infinite stream of random numbers from
+//! a small seed. There exist algorithms that produce random numbers that are in
+//! no measurable way any less random than pulling values 'out of thin air'.
+//!
+//! There are two very different groups interested in developing PRNGs: one
+//! interested in simulations and statistics, another in cryptography. From both
+//! groups Rand picks one good PRNG algorithm. They are exposed as:
+//!
+//! - [`SmallRng`]. Uses little memory, and is very fast.
+//!   Note: The currently picked algorithm is not ideal, but in a future version
+//!   the values will for all practical purposes be statistically indiscernible
+//!   from true randomness.
+//! - [`StdRng`]. In addition to the output being indiscernible from true
+//!   randomness, this PRNG is unpredictable, even if you know the
+//!   implementation. This requirement makes it slower and/or use more memory
+//!   however.
+//!
+//! One of the problems with PRNGs is that they need a seed (which is what the
+//! [`SeedableRng`] trait is for). Where do you get such a seed? Two options:
+//!
+//! - Just use a constant. This is good when you need reproducible results, for
+//!   example to procedurally generate a game world). See the
+//!   [`SeedableRng::from_seed`] documentation.
+//! - Get a random value from somewhere at runtime. The
+//!   [`FromEntropy::from_entropy`] method helps here. It will usually get the
+//!   seed from the operating system's PRNG (using [`EntropyRng`]).
+//!
+//! Another problem is that the internal state of a PRNG must change after every
+//! generated number. So an PRNG always has to be mutable. For APIs this means
+//! they always need to somehow preserve the state between function calls, and
+//! that there can always be only one mutable reference to the PRNG.
+//!
+//! [`ThreadRng`] abstracts away most of these issues. It uses [`StdRng`], which
+//! is always a safe choice. Seeding happens automatically on first use. The
+//! PRNG its state is stored in thread local storage, so no need to plumb the
+//! state through an API. You can easily get a mutable reference to it with the
+//! [`thread_rng`] function.
+//!
+//! ### Conclusion
+//!
+//! - [`thread_rng`] is what you often want to use.
+//! - If you want more control or better performance, use [`StdRng`] or
+//!   [`SmallRng`].
```

Comment:
`SmallRng` is not always faster — maybe emphasise benchmarking? Also for more control one should specify the algorithm. I think the main use-case for specifying `StdRng` is going to be in generic code.